### PR TITLE
LinalgRefactor - Convolve

### PR DESCRIPTION
**Pull request by OXPHOS**

### CODE REVIEW

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h [hunk 1]
```diff
@@ -294,6 +304,48 @@ class LinalgBackendEigen : public LinalgBackendBase
 		result_eig = alpha * a_eig + beta * b_eig;
 	}
 
+	/** Eigen3 convolve/cross-correlation method. */
+	template <typename T>
+	void convolve_impl(SGMatrix<T>& X, SGMatrix<T>& W, SGMatrix<T>& Y,
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: Shouldn't these be const references?

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h [hunk 2]
```diff
@@ -294,6 +304,48 @@ class LinalgBackendEigen : public LinalgBackendBase
...
+			{
+				int32_t yout = y/stride_y;
+
+				T sum = 0;
+				for (int32_t x1=x-rx; x1<=x+rx; x1++)
+				{
+					int32_t wx = flip ? x1-x+rx : rx-x1+x;
+					for (int32_t y1=y-ry; y1<=y+ry; y1++)
+					{
+						if (x1>=0 && y1>=0 && x1<width && y1<height)
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: This doesn't look like eigen3 code

Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: Also this really badly needs multicore acceleration 
Openmp, did you ever use that?

File: src/shogun/mathematics/linalg/LinalgBackendViennacl.h [hunk 3]
```diff
@@ -262,6 +272,46 @@ class LinalgBackendViennaCL : public LinalgBackendGPUBase
...
+			kernel.global_work_size(0, linalg::implementation::ocl::align_to_multiple_2d(Y.num_cols));
+			kernel.global_work_size(1, linalg::implementation::ocl::align_to_multiple_2d(Y.num_rows));
+
+			viennacl::ocl::enqueue(kernel(
+				X_gpu->data_matrix(X.num_rows, X.num_cols), cl_int(X.num_cols),
+				cl_int(X.num_rows), cl_int(X_gpu->m_offset),
+				W_gpu->data_matrix(W.num_rows, W.num_cols), cl_int(W_gpu->m_offset),
+				Y_gpu->data_matrix(Y.num_rows, Y.num_cols), cl_int(Y_gpu->m_offset)));
+		}
+		else
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendViennacl.h]: Really hard to read. Is this tested?

File: src/shogun/mathematics/linalg/LinalgBackendViennaclKernels.h [hunk 4]
```diff
@@ -41,6 +41,204 @@
...
+						for (int rx=0; rx<=2*RADIUS_X; rx++)
+							for (int ry=0; ry<=2*RADIUS_Y; ry++)
+								X_local[xl+rx][yl+ry] = readX(X, x-RADIUS_X+rx, y-RADIUS_Y+ry, X_width, X_height, X_offset);
+					}
+					else if (xl==WORK_GROUP_SIZE_2D-1)
+					{
+						for (int rx=0; rx<=2*RADIUS_X; rx++)
+							X_local[xl+rx][yl] = readX(X, x-RADIUS_X+rx, y-RADIUS_Y, X_width, X_height, X_offset);
+					}
+					else if (yl == WORK_GROUP_SIZE_2D-1)
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendViennaclKernels.h]: This is crazy code
Did the original state where it's coming from? Is it tested? Used?

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 5]
```diff
@@ -294,6 +294,54 @@ void add(SGMatrix<T>& a, SGMatrix<T>& b, SGMatrix<T>& result, T alpha=1, T beta=
 	infer_backend(a, b)->add(a, b, alpha, beta, result);
 }
 
+/** Computes the 2D convolution of X with W
+ * This version returns the result in a newly created matrix.
+ *
+ * @param X Input image
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: Just say matrix, not image

Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: Goes everywhere

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 6]
```diff
@@ -294,6 +294,54 @@ void add(SGMatrix<T>& a, SGMatrix<T>& b, SGMatrix<T>& result, T alpha=1, T beta=
...
+ * @param flip If true the filter coefficients are flipped, performing cross-correlation
+ *  instead of convolution
+ * @param stride_x Stride in the x (column) direction
+ * @param stride_y Stride in the y (row) direction
+ */
+template <typename T>
+SGMatrix<T> convolve(SGMatrix<T>& X, SGMatrix<T>& W, bool flip=false,
+			index_t stride_x=1, index_t stride_y=1)
+{
+	SGMatrix<T> result(X.num_rows/stride_y, X.num_cols/stride_x);
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: I like these wrapped methods and the possibility to pass preallicated mem :)

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h (hunk 2 above)
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: This is not eigen3 and I didn't find `convolve` method in eigen. 
Unfortunately I haven't used Openmp

File: src/shogun/mathematics/linalg/LinalgBackendViennacl.h (hunk 3 above)
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendViennacl.h]: Yes they are tested. I think the problem is the code `X.vcl_matrix()` is now like `X_gpu->data_matrix(X.num_rows, X.num_cols)`

File: src/shogun/mathematics/linalg/LinalgBackendViennaclKernels.h (hunk 4 above)
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendViennaclKernels.h]: Nope..They're not tested on my machine but there were ViennaCL convolution tests in unit-tests. I found only one application of convolution in shogun (see [here](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/neuralnets/ConvolutionalFeatureMap.cpp#L267)) but it actually defines its own method (the same as the "linalg<Eigen>" one). So in a word the `linalg::convolve` method is not used.

File: src/shogun/mathematics/linalg/LinalgBackendEigen.h (hunk 2 above)
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendEigen.h]: grep for convlution here:
http://eigen.tuxfamily.org/index.php?title=3.3

We can take code from there, this one should not be in, it is slow and unreadable. I know it is not your fault, but if we have unit tests it should be easy to update the code.....

File: src/shogun/mathematics/linalg/LinalgBackendViennacl.h (hunk 3 above)
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgBackendViennacl.h]: ok, so we need to make sure we test this properly after a merge

File: src/shogun/mathematics/linalg/LinalgBackendViennaclKernels.h (hunk 4 above)
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/LinalgBackendViennaclKernels.h]: Agree! Lemme remove the old codes in Shogun

**Comment by OXPHOS** [comment]
@karlnapf updated

**Comment by karlnapf** [comment]
Good. Better no code than shit code !

**Pull request merged**